sensor_msgs/include/sensor_msgs/point_cloud_msg_wrapper.hpp

```diff
@@ -0,0 +1,496 @@
+    reset_msg(m_cloud_ref.header.frame_id);
+  }
+
+  /// @brief      Allocate memory to hold a specified number of points.
+  COMPILE_IF_MUTABLE(CloudMsgT, void) reserve(const size_t expected_number_of_points) {
+    m_cloud_ref.data.reserve(m_cloud_ref.point_step * expected_number_of_points);
+  }
+
+  /// A constant iterator to the beginning of data.
+  ConstIterator begin() const noexcept {return cbegin();}
+  /// A constant iterator to the end of data.
+  ConstIterator end() const noexcept {return cend();}
+  /// A constant iterator to the beginning of data.
+  ConstIterator cbegin() const noexcept {return ConstIterator{this, 0UL};}
+  /// A constant iterator to the end of data.
+  ConstIterator cend() const noexcept {return ConstIterator{this, size()};}
+
+private:
+  /// Allocate additional memory in the end of data field of the message.
+  COMPILE_IF_MUTABLE(CloudMsgT, void) extend_data_by(const std::size_t bytes_to_allocate)
+  {
+    const auto new_size = m_cloud_ref.data.size() + bytes_to_allocate;
+    m_cloud_ref.data.resize(new_size);
+  }
+
+  /// @brief      Reset the message fields to match the members of the PointT struct. The point
+  ///             cloud message is ready for modification after this operation.
+  COMPILE_IF_MUTABLE(CloudMsgT, void) reset_msg(
+    const FieldNameT & frame_id,
+    const sensor_msgs::msg::PointCloud2::_fields_type & generated_fields)
+  {
+    // TODO(igor): All these settings can also be specified through some input struct if needed in
+    // the future. This is omitted for now.
+    m_cloud_ref.fields = generated_fields;
+    m_cloud_ref.height = 1U;
+    m_cloud_ref.is_bigendian = false;
+    m_cloud_ref.is_dense = false;
+    m_cloud_ref.header.frame_id = frame_id;
+    m_cloud_ref.point_step = sizeof(PointT);
+    m_cloud_ref.data.clear();
+    m_cloud_ref.width = 0U;
+    m_cloud_ref.row_step = 0U;
+  }
+
+  bool check_that_generated_fields_cover_all_point_members(
+    const sensor_msgs::msg::PointCloud2::_fields_type & generated_fields)
+  {
+    std::int32_t sum_of_all_fields_sizes{};
+    for (const auto & field : generated_fields) {
+      sum_of_all_fields_sizes +=
+        static_cast<std::int32_t>(field.count) * sizeof_field(field.datatype);
+    }
+    if (sum_of_all_fields_sizes == sizeof(PointT)) {
+      // The generated fields cover all members of the struct without gaps.
+      return true;
+    }
+    // The idea is the following:
+    // 1. Allocate storage as big as a point.
+    // 2. Fill the storage with zeros.
+    // 3. Iterate through all fields and set the appropriate memory to ones.
+    // 4. Flip all bits - now, if all fields are present, only the padding is non-zero.
+    // 5. Reinterpret this memory as a point.
+    // 6. Compare this point to a value initialized one.
+    // If these points are equal - we have all fields. Otherwise, we are missing a field.
+    alignas(PointT) std::uint8_t buffer[sizeof(PointT)];
```

Comment:
Just to start a discussion: this is one of the most strange parts in this code. Basically, this is invented for the purposes of this class and I am not aware of other usages just like this one. Why? Because we need to make sure that we cover all the fields in case there is a point struct that has field that have no corresponding generators passed into this class. Any ideas on how to make this simpler or how we can get rid of this are welcome. Otherwise, any suggestions on how to test this even better are also much appreciated.

Reply:
The current state of this code only checks if the sizes of fields match the size of fields in the given PointType struct right?

Because flipping bits of a 0(uint32_t) is same as flipping bits 2 x 0(uint16_t)s.

Like if PointType had:
uint32_t x;

and fields were like:
- uint16_t x
- uint16_t y

It would pass this test as correct even though they don't match right?

Reply:
Also following for loop in the code would give a segfault if the fields were accounting for more memory space than the PointType:

Example if 
- fields had x,y,z,intensity for 4x4=16 bytes
- PointType had only x,y,z for 4x3=12 bytes (assume no padding)

```cpp
    alignas(PointT) std::uint8_t buffer[sizeof(PointT)];
    std::fill(buffer, buffer + sizeof(PointT), 0U);
    for (const auto & field : generated_fields) {
      const auto field_first_byte = field.offset;
      const auto field_last_byte =
        field.offset + field.count * static_cast<std::size_t>(sizeof_field(field.datatype));
      std::fill(
        &buffer[field_first_byte],
        &buffer[field_last_byte],
        std::numeric_limits<std::uint8_t>::max());
    }
```

here the `std::fill` will try to access 13th element of buffer which has size 12 and segfault.

Reply:
> It would pass this test as correct even though they don't match right?

this particular part - yes, but I do (at least I hope) all the checks that account for these situations before this check starts.

>  the code would give a segfault if the fields were accounting for more memory space than the PointType

I don't think it will. The generated fields are generated _from the same `PointType`_. These are not the fields that are provided by the user in any way. 

You can try to reproduce the failures you suggest by adding a certain point type to the test set. Do the ones you suggest fail the test?

Reply:
I have added the test to here:

https://github.com/xmfcx/common_interfaces/commit/9cfa2376d4fda69ec0af81ead41086126a294baa

and it fails with following output:
```cpp
PointCloudMsgWrapperTest.point_with_2_custom_fields
30: /home/mfc/projects/ros2_ws/src/common_interfaces/sensor_msgs/test/test_point_cloud_msg_wrapper.cpp:283: Failure
30: Expected: CustomCloudModifier(msg, "some_frame_id") doesn't throw an exception.
30:   Actual: it throws.
30: unknown file: Failure
30: C++ exception with description "Generated fields don't match the members of the point struct. Make sure you have correct generators tuple passed into this class." thrown in the test body.
30: [  FAILED  ]
```

I'm not sure what is the best way to link my changes here since I cannot push into your remote and you cannot push into mine.

Reply:
@xmfcx I in the code you linked you are using a different point structure (`PointWithCustomField`) from the one you define. So for now it makes sense that it fails. Please change it to `PointCustomAB` and check again.

I think what we could do is you could open a PR into my branch, this way I can merge your changes directly into this branch. Would that work for you?

Reply:
Also do you think [BOOST_FUSION_ADAPT_STRUCT](https://www.boost.org/doc/libs/1_68_0/libs/fusion/doc/html/fusion/adapted/adapt_struct.html) could be used here as suggested in [here](https://stackoverflow.com/a/17660354)?

Not sure if we are allowed to use boost here but it would make generating the fields less macro-y for this code.
More like it would offload the macro parts to boost and we could just use `boost::fusion::for_each()` to iterate through the fields of the point struct and generate fields on the runtime. 
And we wouldn't need user entering anything to generate the fields.

Reply:
> @xmfcx I in the code you linked you are using a different point structure (`PointWithCustomField`) from the one you define. So for now it makes sense that it fails. Please change it to `PointCustomAB` and check again.

Ah my bad, sorry :P
Fixed it and now works.

> I think what we could do is you could open a PR into my branch, this way I can merge your changes directly into this branch. Would that work for you?

Ok, made the PR.

Reply:
I am not too familiar with boost fusion. I'm not sure what is the policy of Boost usage here, so I will not target this in my comment. I do not think we can directly use it here though as I just don't see how this could be implemented without an input from the user in the form of generators or alike. For it to work without user input we would need reflection-like capabilities and C++ does not have that unfortunately. But if you know of a way to generate the fields entries from a random struct at runtime - I am all ears.

Reply:
Thinking more about this, it might make sense to do smth _like_ boost does. Basically, we would wrap every struct that the user might provide (or instruct the user to do it on their own) and pass the resulting iterable object to our class where we can generate the fields from it etc. The difference here is that the user would need to do it for every type that they might want to use. The up side is that the interface seems to be quite clean. I might have a deeper look into this.

Reply:
We have a policy of no boost in the core, so we shouldn't make any of this dependent on boost.  Alternatively, if you find you really need to use boost, you can make this code a third-party package that depends on boost.

Reply:
I was thinking if we could use `BOOST_FUSION_ADAPT_STRUCT` in the template typename the wrapper takes or something. But with no-boost policy, it can't be helped.

Reply:
@christophebedard yep, that is exactly what I expected, thanks for pitching in!

Reply:
@dejanpan oh no! Sorry @christophebedard I meant to mention @clalancette here :man_facepalming:  Sorry for the confusion guys.